core/domain/exp_services.py

```diff
@@ -1958,16 +2147,21 @@ def are_changes_mergeable(exp_id, frontend_version, change_list):
                 if change.state_name in state_names_of_renamed_states:
                     old_state_name = (
                         state_names_of_renamed_states[change.state_name])
-                    new_state_name = (
-                        state_names_of_renamed_states[change.state_name])
-                if change.state_name in old_to_new_state_names:
-                    new_state_name = old_to_new_state_names[old_state_name]
+                if old_state_name in old_to_new_state_names:
+                    if feconf.CAN_SEND_EMAILS:
+                        email_manager.send_mail_to_admin(
+                            'Adding/Deleting State Changes not mergeable',
+                            'The exploration id is: %s, \n '
+                            'The frontend version is: %s \n '
+                            'The changes list is: %s '
+                            % (exp_id, frontend_version, change_list))
+                    return False
                 if old_state_name not in changed_translations:
                     changed_translations[old_state_name] = []
                 frontend_exp_states = (
                     frontend_version_exploration.states[old_state_name])
                 backend_exp_states = (
-                    backend_version_exploration.states[new_state_name])
+                    backend_version_exploration.states[old_state_name])
```

Comment:
Can you explain this change please?

Reply:
Actually, that was an error. Here new_state_name use to refer to the changed state name in the latest version after the rename changes applied. But here backend_version_exploration is the exploration before the change list applied so that changed state name doesn't exist here. So fixed this error and used old_state_name.

Reply:
This was merged in your previous PR, right? The fact that this got merged probably means that the backend tests don't handle all edge cases. Make sure to handle all edge cases as well in backend tests cases and not just go for coverage. Especially for this function, since it's quite a large and complicated function.
@kevintab95 Can you make sure all edge cases are handled?

Reply:
@aks681 Yeah, this was merged in my previous PR. But the error wasn't that I used the wrong value here. The error was related to handling the state name renames encountered after completing the functionality during the manual testing. The error was related to the frontend where the exploration diff service couldn't handle the state renames if someone opened the editor with an existing unsaved draft after a long time and there is a state rename in the backend. So when someone opens the editor with an existing unsaved draft, the editor contains the latest version of an exploration. Therefore if there were any state renaming in the backend, the editor includes the states with the latest names but the unsaved draft change list still contains the changes according to the old state names. So if we wanted to merge it in the backend, it was possible to handle state renames cases. Still, when we try to save them, an error was thrown by the exploration diff service because it could not show the changes because of the different state names as there is no tracking of states in the frontend. So after looking into the codebase to try to handle these cases and after discussing it with @kevintab95, we decided that for now, we'll discard the changes if any state is renamed in the backend and the changes are made in that state in the frontend, and we'll inform the admin about this scenario so that these cases can be reviewed and appropriately handled. 
So in my previous PR, we were handling the state renames; hence there were two variables, first is the old_state_name, which used to store the state name from the frontend_version, and the second was new_state_name which used to store the latest name of a state after multiple renames in the backend version. But after we stopped tracking the backend version state rename changes, we no longer needed new_state_name, and as both new_state_name and old_state_name are the same now, I entirely removed new_state_name and replaced it with the old_state_name.